Python/traceback.c

```diff
@@ -512,8 +516,147 @@ _Py_DisplaySourceLine(PyObject *f, PyObject *filename, int lineno, int indent, i
+ *    File "/home/isidentical/cpython/cpython/t.py", line 10, in <module>
+ *      add_values(1, 2, 'x', 3, 4)
+ *      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+ *    File "/home/isidentical/cpython/cpython/t.py", line 2, in add_values
+ *      return a + b + c + d + e
+ *             ~~~~~~^~~
+ *  TypeError: 'NoneType' object is not subscriptable
+ */
+
+#define IS_WHITESPACE(c) (((c) == ' ') || ((c) == '\t') || ((c) == '\f'))
+
+static int
+extract_anchors_from_expr(PyObject *segment, expr_ty expr, int *left_anchor, int *right_anchor)
+{
+    switch (expr->kind) {
+        case BinOp_kind: {
+            PyObject *operator = PyUnicode_Substring(segment, expr->v.BinOp.left->end_col_offset,
+                                                     expr->v.BinOp.right->col_offset);
+            if (!operator) {
+                return -1;
+            }
+
+            const char *operator_str = PyUnicode_AsUTF8(operator);
+            if (!operator_str) {
+                Py_DECREF(operator);
+                return -1;
+            }
+
+            Py_ssize_t i, len = PyUnicode_GET_LENGTH(operator);
+            for (i = 0; i < len; i++) {
+                if (IS_WHITESPACE(operator_str[i])) {
+                    continue;
+                }
+
+                int index = Py_SAFE_DOWNCAST(i, Py_ssize_t, int);
+                *left_anchor = expr->v.BinOp.left->end_col_offset + index;
+                *right_anchor = expr->v.BinOp.left->end_col_offset + index + 1;
+                if (i + 1 < len && !IS_WHITESPACE(operator_str[i + 1])) {
+                    ++*right_anchor;
+                }
+                break;
+            }
+            Py_DECREF(operator);
+            return 1;
+        }
+        case Subscript_kind: {
+            *left_anchor = expr->v.Subscript.value->end_col_offset;
+            *right_anchor = expr->v.Subscript.slice->end_col_offset + 1;
+            return 1;
+        }
+        default:
+            return 0;
+    }
+}
+
+static int
+extract_anchors_from_stmt(PyObject *segment, stmt_ty statement, int *left_anchor, int *right_anchor)
+{
+    switch (statement->kind) {
+        case Expr_kind: {
+            return extract_anchors_from_expr(segment, statement->v.Expr.value, left_anchor, right_anchor);
+        }
+        default:
+            return 0;
+    }
+}
+
+static int
+extract_anchors_from_line(PyObject *filename, PyObject *line,
+                          Py_ssize_t start_offset, Py_ssize_t end_offset,
+                          int *left_anchor, int *right_anchor)
+{
+    int res = -1;
+    PyArena *arena = NULL;
+    PyObject *segment = PyUnicode_Substring(line, start_offset, end_offset);
+    if (!segment) {
+        goto done;
+    }
+
+    const char *segment_str = PyUnicode_AsUTF8(segment);
+    if (!segment) {
+        goto done;
+    }
+
+    arena = _PyArena_New();
+    if (!arena) {
+        goto done;
+    }
+
+    PyCompilerFlags flags = _PyCompilerFlags_INIT;
+
+    _PyASTOptimizeState state;
```

Comment:
There's some fields in `_PyASTOptimizeState` here that will be unset like `recursion_depth` and `recursion_limit`. Is that safe? 

Also, maybe try to use struct initialization syntax, something like this?

```c
_PyASTOptimizeState state = {
    .optimize = _Py_GetConfig()->optimization_level,
    .ff_features = 0,
    .recursion_depth = ?,
    .recursion_limit = ?
};
```

Reply:
These are set by optimized itself;
https://github.com/python/cpython/blob/248173cc0483a9ad9261353302f1234cf9eb2ebe/Python/ast_opt.c#L1103-L1105

Also see how it is been used so far;
https://github.com/python/cpython/blob/248173cc0483a9ad9261353302f1234cf9eb2ebe/Python/compile.c#L465-L471